date: `date -d0` results to a different output

Hello,

First, thank you for maintaining and developing this ambitious project!

TL;DR: the output of `date -d0` is different from the GNU version, which is different from the busybox version.
- ```
  coreutils date -d0
  Fri Dec 31 22:17:30 LMT -0001
  ```
- ```
  /usr/bin/gnudate -d0
  Mon 13 Oct 00:00:00 CEST 2025
  ```
- ```
  busybox date -d0
  date: invalid date '0'
  ```

To be honest, when looking at the man page, I don't know who is right and who is wrong. But the [GNU manual](https://www.gnu.org/software/coreutils/manual/html_node/Pure-numbers-in-date-strings.html) describes this behaviour, so I guess the Rust version should do the same, and parse decimal numbers of the form `hhmm` or just `hh` as hours and minutes from the current day, no?

(Note: With the Rust version, I don't know why we have this output, but the `+%s` version gives `-62167226399`. FYI, when I feed this to the GNU version (`/usr/bin/gnudate -d"@-62167226399"`), we get **almost** the same output: `Fri 31 Dec 22:17:31 LMT -001`, one-second difference ; is it normal? Should I create another issue)

---------------------------

On my side, the exact value for this doesn't really matter, but when using the Rust's date version, I have issues when building the Linux kernel:

```
$ KBUILD_BUILD_TIMESTAMP="0" make
(...)
  ERROR: Timestamp out of range for cpio format
  make[4]: *** [usr/Makefile:76: usr/initramfs_data.cpio] Error 1
  make[4]: *** Deleting file 'usr/initramfs_data.cpio'
  make[3]: *** [scripts/Makefile.build:556: usr] Error 2
  make[3]: *** Waiting for unfinished jobs....
```

I have this issue, because [`date -d"0" +%s`](https://github.com/torvalds/linux/blob/3a8660878839faadb4f1a6dd72c3179c1df56787/usr/gen_initramfs.sh#L225) is used, and returns an [unsupported negative value](https://github.com/torvalds/linux/blob/3a8660878839faadb4f1a6dd72c3179c1df56787/usr/gen_init_cpio.c#L701-L709).

There are different workarounds (`KBUILD_BUILD_TIMESTAMP="-1"` or disabling `CONFIG_BLK_DEV_INITRD`), but I guess the issue should be fixed on uutils' coretutils side, and not adding a special case in the Linux kernel build system to avoid negative values, right?

--- a/src/uu/date/src/date.rs
+++ b/src/uu/date/src/date.rs
@@ -203,7 +203,48 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
     // Iterate over all dates - whether it's a single date or a file.
     let dates: Box<dyn Iterator<Item = _>> = match settings.date_source {
         DateSource::Human(ref input) => {
-            let date = parse_date(input);
+            // GNU compatibility (Pure numbers in date strings):
+            // - Manual: https://www.gnu.org/software/coreutils/manual/html_node/Pure-numbers-in-date-strings.html
+            // - Semantics: a pure decimal number denotes today’s time-of-day (HH or HHMM).
+            //   Examples: "0"/"00" => 00:00 today; "7"/"07" => 07:00 today; "0700" => 07:00 today.
+            // For all other forms, fall back to the general parser.
+            let is_pure_digits =
+                !input.is_empty() && input.len() <= 4 && input.chars().all(|c| c.is_ascii_digit());
+
+            let date = if is_pure_digits {
+                // Derive HH and MM from the input
+                let (hh_opt, mm_opt) = if input.len() <= 2 {
+                    (input.parse::<u32>().ok(), Some(0u32))
+                } else {
+                    let (h, m) = input.split_at(input.len() - 2);
+                    (h.parse::<u32>().ok(), m.parse::<u32>().ok())
+                };
+
+                if let (Some(hh), Some(mm)) = (hh_opt, mm_opt) {
+                    // Compose a concrete datetime string for today with zone offset.
+                    // Use the already-determined 'now' and settings.utc to select offset.
+                    let date_part =
+                        strtime::format("%F", &now).unwrap_or_else(|_| String::from("1970-01-01"));
+                    // If -u, force +00:00; otherwise use the local offset of 'now'.
+                    let offset = if settings.utc {
+                        String::from("+00:00")
+                    } else {
+                        strtime::format("%:z", &now).unwrap_or_default()
+                    };
+                    let composed = if offset.is_empty() {
+                        format!("{date_part} {hh:02}:{mm:02}")
+                    } else {
+                        format!("{date_part} {hh:02}:{mm:02} {offset}")
+                    };
+                    parse_date(composed)
+                } else {
+                    // Fallback on parse failure of digits
+                    parse_date(input)
+                }
+            } else {
+                parse_date(input)
+            };
+
             let iter = std::iter::once(date);
             Box::new(iter)
         }

--- a/tests/by-util/test_date.rs
+++ b/tests/by-util/test_date.rs
@@ -778,3 +778,60 @@ fn test_date_resolution_no_combine() {
         .arg("2025-01-01")
         .fails();
 }
+
+#[test]
+fn test_date_numeric_d_basic_utc() {
+    // Verify GNU-compatible pure-digit parsing for -d STRING under UTC
+    // 0/00 -> today at 00:00; 7/07 -> today at 07:00; 0700 -> today at 07:00
+    let today = Utc::now().date_naive();
+    let yyyy = today.year();
+    let mm = today.month();
+    let dd = today.day();
+
+    let mk =
+        |h: u32, m: u32| -> String { format!("{yyyy:04}-{mm:02}-{dd:02} {h:02}:{m:02}:00 UTC\n") };
+
+    new_ucmd!()
+        .env("TZ", "UTC0")
+        .arg("-d")
+        .arg("0")
+        .arg("+%F %T %Z")
+        .succeeds()
+        .stdout_only(mk(0, 0));
+
+    new_ucmd!()
+        .env("TZ", "UTC0")
+        .arg("-d")
+        .arg("7")
+        .arg("+%F %T %Z")
+        .succeeds()
+        .stdout_only(mk(7, 0));
+
+    new_ucmd!()
+        .env("TZ", "UTC0")
+        .arg("-d")
+        .arg("0700")
+        .arg("+%F %T %Z")
+        .succeeds()
+        .stdout_only(mk(7, 0));
+}
+
+#[test]
+fn test_date_numeric_d_invalid_numbers() {
+    // Ensure invalid HHMM values are rejected (GNU-compatible)
+    new_ucmd!()
+        .env("TZ", "UTC0")
+        .arg("-d")
+        .arg("2400")
+        .arg("+%F %T %Z")
+        .fails()
+        .stderr_contains("invalid date");
+
+    new_ucmd!()
+        .env("TZ", "UTC0")
+        .arg("-d")
+        .arg("2360")
+        .arg("+%F %T %Z")
+        .fails()
+        .stderr_contains("invalid date");
+}

